### PR TITLE
pkcs12: initial types with decoders/encoders

### PR DESCRIPTION
**Pull request by carl-wallace**
- Encoders and decoders only for PKCS 12 (decryption support and a builder will be submitted subsequently)
- Does not support sdsiCerts or alternative CRL types
- Copies the Pbkdf2Params and EncryptedPrivateKeyInfo definitions to avoid the lifetime requirement of the definitions in the pkcs5 crate
- Copies the Pbes2Params definition to avoid EncryptionScheme requirement of the definition in the pkcs5 crate
- Copies the PrivateKeyInfo definition to avoid the lifetime requirement of the definition in the pkcs8 crate
- SafeBag is a bit unusual owing to the OR'ed list of options (this will largely be hidden behind decryption and builder support)

### CODE REVIEW

**Comment by tarcieri** [comment]
@carl-wallace can you rebase?

**Comment by carl-wallace** [comment]
I can do that first thing tomorrow.

File: pkcs12/src/lib.rs [hunk 1]
```diff
@@ -1,3 +1,5 @@
+//! TODO: complete PKCS#12 crate
+
```
Comment (tarcieri) [inline pkcs12/src/lib.rs:2]: This can probably be removed now

```suggestion
```

File: pkcs12/src/safe_bag.rs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+#[allow(missing_docs)]
+pub struct SafeBag {
+    pub bag_id: ObjectIdentifier,
+    //#[asn1(context_specific = "0", tag_mode = "EXPLICIT")]
+    pub bag_value: Vec<u8>,
+    pub bag_attributes: Option<Attributes>,
+}
+
+impl<'__der_lifetime> ::der::DecodeValue<'__der_lifetime> for SafeBag {
+    fn decode_value<R: ::der::Reader<'__der_lifetime>>(
```
Comment (tarcieri) [inline pkcs12/src/safe_bag.rs:40]:
```suggestion
impl<'a> ::der::DecodeValue<'a> for SafeBag {
    fn decode_value<R: ::der::Reader<'a>>(
```

File: pkcs12/src/safe_bag.rs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+///     bagAttributes  SET OF PKCS12Attribute OPTIONAL
+/// }
+/// ```
+///
+/// [RFC 7292 Section 4]: https://www.rfc-editor.org/rfc/rfc7292#section-4.2
+#[derive(Clone, Debug, Eq, PartialEq)]
+#[allow(missing_docs)]
+pub struct SafeBag {
+    pub bag_id: ObjectIdentifier,
+    //#[asn1(context_specific = "0", tag_mode = "EXPLICIT")]
```
Comment (tarcieri) [inline pkcs12/src/safe_bag.rs:34]: What's wrong with derive here?

File: pkcs12/src/safe_bag.rs (hunk 2 above)
Comment (carl-wallace) [reply pkcs12/src/safe_bag.rs]: changed

File: pkcs12/src/lib.rs [hunk 4]
```diff
@@ -14,10 +16,84 @@
...
+// cert types
+/// `pkcs-9 x509Certificate for pkcs-12` Object Identifier (OID).
+pub const PKCS_12_X509_CERT_OID: ObjectIdentifier =
+    ObjectIdentifier::new_unwrap("1.2.840.113549.1.9.22.1");
+
+/// `pkcs-9 sdsiCertificate for pkcs-12` Object Identifier (OID).
+pub const PKCS_12_SDSI_CERT_OID: ObjectIdentifier =
+    ObjectIdentifier::new_unwrap("1.2.840.113549.1.9.22.2");
+
+// todo: return the friendly name if present? (minimally, defer until BMPString support is available)
```
[the remark this answers is not in the document]
Comment (carl-wallace) [reply pkcs12/src/lib.rs:92]: OK. I will take a whack at supporting friendly name via the new BMPString support and will move the remaining todos to an issue once all review items all done.

[the remark this answers is not in the document]
Comment (tarcieri) [reply pkcs12/src/lib.rs:92]: I can go ahead and merge with this if you'd like to follow up

**Pull request merged**